server/src/internalClusterTest/java/org/elasticsearch/rest/ChunkedZipResponseIT.java

```diff
@@ -0,0 +1,500 @@
+                assertEquals("application/zip", response.getHeaders("Content-Type")[0].getValue());
+                final var contentDispositionHeader = response.getHeaders("Content-Disposition")[0].getElements()[0];
+                assertEquals("attachment", contentDispositionHeader.getName());
+                assertEquals(
+                    RandomZipResponsePlugin.RESPONSE_FILENAME + ".zip",
+                    contentDispositionHeader.getParameterByName("filename").getValue()
+                );
+                responseStarted.countDown();
+            }
+
+            @Override
+            public void consumeContent(ContentDecoder decoder, IOControl ioControl) throws IOException {
+                readBuffer.clear();
+                final var bytesRead = decoder.read(readBuffer);
+                if (bytesRead > 0) {
+                    bytesToConsume -= bytesRead;
+                }
+
+                if (bytesToConsume <= 0) {
+                    bodyConsumed.countDown();
+                    ioControl.shutdown();
+                }
+            }
+
+            @Override
+            public void responseCompleted(HttpContext context) {}
+
+            @Override
+            public void failed(Exception ex) {}
+
+            @Override
+            public Exception getException() {
+                return null;
+            }
+
+            @Override
+            public HttpResponse getResult() {
+                return null;
+            }
+
+            @Override
+            public boolean isDone() {
+                return false;
+            }
+
+            @Override
+            public void close() {}
+
+            @Override
+            public boolean cancel() {
+                return false;
+            }
+        }));
+
+        try (var restClient = createRestClient(internalCluster().getRandomNodeName())) {
+            // one-node REST client to avoid retries
+            expectThrows(ConnectionClosedException.class, () -> restClient.performRequest(request));
+        }
+        safeAwait(responseStarted);
+        safeAwait(bodyConsumed);
+        assertNull(getExpectedEntries()); // mainly just checking that all refs are released
+    }
+
+    public void testGetNextPartFailure() throws IOException {
+        final var request = new Request("GET", RandomZipResponsePlugin.INFINITE_ROUTE);
+        request.addParameter(RandomZipResponsePlugin.GET_NEXT_PART_COUNT_DOWN_PARAM, Integer.toString(between(1, 100)));
+
+        try (var restClient = createRestClient(internalCluster().getRandomNodeName())) {
+            // one-node REST client to avoid retries
+            assertThat(
+                safeAwaitFailure(
+                    Response.class,
+                    l -> restClient.performRequestAsync(request, ActionTestUtils.wrapAsRestResponseListener(l))
+                ),
+                anyOf(instanceOf(ConnectionClosedException.class), instanceOf(MalformedChunkCodingException.class))
```

Comment:
For my education: Can it fail with `MalformedChunkCodingException`? It seems to me that the connection is closed in both cases which results in `ConnectionClosedException`?

Reply:
I agree it's surprising but yeah this is something the REST client will return indeed. Try removing it from this line and run the test a few hundred times.